packages/ra-core/src/auth/useAuthState.ts

```diff
@@ -52,43 +52,64 @@ const emptyParams = {};
 const useAuthState = (
     params: any = emptyParams,
     logoutOnFailure: boolean = false,
-    queryOptions?: UseQueryOptions<boolean, any>
+    queryOptions: UseAuthStateOptions = emptyParams
 ): State => {
     const authProvider = useAuthProvider();
     const logout = useLogout();
     const basename = useBasename();
     const notify = useNotify();
+    const { onSuccess, onError, ...options } = queryOptions;
 
-    const result = useQuery<boolean, any>(
-        ['auth', 'checkAuth', params],
-        () => {
+    const result = useQuery<boolean, any>({
+        queryKey: ['auth', 'checkAuth', params],
+        queryFn: () => {
             // The authProvider is optional in react-admin
-            return authProvider?.checkAuth(params).then(() => true);
+            if (!authProvider) {
+                return true;
+            }
+            return authProvider.checkAuth(params).then(() => true);
         },
-        {
-            onError: error => {
-                const loginUrl = removeDoubleSlashes(
-                    `${basename}/${defaultAuthParams.loginUrl}`
+        retry: false,
+        ...options,
+    });
+
+    useEffect(() => {
+        if (result.data != null && onSuccess) {
+            onSuccess(result.data);
+        }
+    }, [onSuccess, result.data]);
+
+    useEffect(() => {
+        if (result.error != null) {
+            if (onError) {
+                return onError(result.error);
+            }
+
+            const loginUrl = removeDoubleSlashes(
+                `${basename}/${defaultAuthParams.loginUrl}`
+            );
+            if (logoutOnFailure) {
+                logout(
+                    {},
+                    result.error && result.error.redirectTo != null
+                        ? result.error.redirectTo
+                        : loginUrl
                 );
-                if (logoutOnFailure) {
-                    logout(
-                        {},
-                        error && error.redirectTo != null
-                            ? error.redirectTo
-                            : loginUrl
+                const shouldSkipNotify =
+                    result.error && result.error.message === false;
+                !shouldSkipNotify &&
+                    notify(
+                        getErrorMessage(
+                            result.error,
+                            'ra.auth.auth_check_error'
+                        ),
+                        {
+                            type: 'error',
+                        }
```

Comment:
Not sure if it's prettier fault, but the previous code was more compact (and therefore easier to read).

Reply:
prettier